src/NHibernate/Hql/Ast/ANTLR/AstPolymorphicProcessor.cs

```diff
@@ -29,8 +29,11 @@ private IASTNode[] Process()
             // Find all the polymorphic query sources
             _nodeMapping = new PolymorphicQuerySourceDetector(_factory).Process(_ast);
 
-            if (_nodeMapping.Count() > 0)
+            if (_nodeMapping.Count > 0)
             {
+                foreach (var kv in _nodeMapping.Where(x => x.Value.Length == 0))
```

Comment:
Shall we throw one exception for all?

Reply:
I still undecided it's a good place for throwing. Also check seems incorrect for polymorphic case as some failed tests show. Seems should be `_nodeMapping.All(x => x.Value.Length == 0)`

Reply:
> Also check seems incorrect for polymorphic case as some failed tests show.

`Can_LazyPropertyNotCastingException` failure reveals that empty database check doesn't work for entities mapped with `polymorphism="explicit"`